src/advanced/timestamping-tutorial.md

```diff
@@ -0,0 +1,929 @@
+   the transaction, its hash and the current time value.
+
+
+```rust
+impl Transaction for TxTimestamp {
+    fn verify(&self) -> bool {
+        self.verify_signature(self.pub_key())
+    }
+
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+        let time = TimeSchema::new(&fork)
+            .time()
+            .get()
+            .expect("Can't get the time");
+
+        let content = self.content();
+        let hash = content.content_hash();
+
+        let mut schema = Schema::new(fork);
+        if let Some(_entry) = schema.timestamps().get(hash) {
+            Err(Error::HashAlreadyExists)?;
+        }
+
+        trace!("Timestamp added: {:?}", self);
+        let entry = TimestampEntry::new(self.content(), &self.hash(), time);
+        schema.add_timestamp(entry);
+        Ok(())
+    }
+}
+```
+
+## Configure API endpoints
+
+Next, we need to implement the node API. We have separated the code that
+configures API endpoints for our Timestamping service into a separate file -
+[src/api.rs][src/api.rs].
+
+First, we need to define the structures required to implement the API endpoints.
+For our Timestamping service, we will define the following three structures:
+
+- the `TimestampQuery` structure which contains only the hash of the
+  timestamping transaction:
+
+  ```rust
+  pub struct TimestampQuery {
+      /// Hash of the requested timestamp.
+      pub hash: Hash,
+  }
+
+  impl TimestampQuery {
+      /// Creates new `TimestampQuery` with given `hash`.
+      pub fn new(hash: Hash) -> Self {
+          TimestampQuery { hash }
+      }
+  }
+  ```
+
+- the `TimestampProof` structure which contains all the data required to prove
+  that a timestamping transaction has been included into the blockchain:
+
+  ```rust
+  pub struct TimestampProof {
+      /// Proof of the last block.
+      pub block_info: BlockProof,
+      /// Actual state hashes of the timestamping service with their proofs.
+      pub state_proof: MapProof<Hash, Hash>,
+      /// Actual state of the timestamping database with proofs.
+      pub timestamp_proof: MapProof<Hash, TimestampEntry>,
+  }
+  ```
+
+- the `PublicApi` structure which enables us to implement public API endpoints
+  for our service:
+
+  ```rust
+  pub struct PublicApi;
+  ```
+
+Now we are ready to define the three endpoints required for our service:
+
+```rust
+pub fn handle_post_transaction(
+        state: &ServiceApiState,
+        transaction: TxTimestamp,
+    ) -> api::Result<Hash> {
+        let hash = transaction.hash();
+        state.sender().send(transaction.into())?;
+        Ok(hash)
+    }
+```
+
+The `handle_post_transaction` function sends a timestamping transaction to the
+blockchain performing the folloiwng operations:
+
+- take a  timestamping transaction
+- calculate its hash
+- send it to the blockchain network, using the methods defined in Exonum
+- add transction event with the timestamping transaction in the form which
```

Comment:
> transction

typo

Reply:
And actual I cannot understand this bullet-point. Previous one is about sending transaction to the network, next action in code is to return transaction hash to user.